### PR TITLE
Implement #1329: Add remove-unused-objects command

### PR DESCRIPTION
**Pull request by oli414**
Implements #1329.
Adds a command to the in-game console to remove unused objects from the object selection.

### CODE REVIEW

**Comment by Margen67** [comment]
Does this work for multiplayer?

**Comment by oli414** [comment]
No, it doesn't. Object selection isn't available in multiplayer

File: src/openrct2/Editor.h [hunk 1]
```diff
@@ -85,6 +85,8 @@ extern "C"
 
     void editor_open_windows_for_current_step();
 
+	sint32 editor_remove_unused_objects();
```
Comment (Gymnasiast) [inline src/openrct2/Editor.h]: Never use tabs. Use four spaces.

File: src/openrct2/interface/console.c [hunk 2]
```diff
@@ -1206,6 +1207,12 @@ static sint32 cc_open(const utf8 **argv, sint32 argc) {
     return 0;
 }
 
+static sint32 cc_remove_unused_objects(const utf8 **argv, sint32 argc) {
```
Comment (Gymnasiast) [inline src/openrct2/interface/console.c]: Braces go on their own lines.

File: src/openrct2/windows/EditorObjectSelection.cpp [hunk 3]
```diff
@@ -1877,3 +1877,41 @@ bool editor_check_object_group_at_least_one_selected(sint32 checkObjectType)
...
     return false;
 }
+
+sint32 editor_remove_unused_objects()
+{
+	bool selectionFlagsExist = _objectSelectionFlags != nullptr;
+	if (!selectionFlagsExist)
+	{
+		if (!sub_6AB211())
+			return 0;
```
Comment (Gymnasiast) [inline src/openrct2/windows/EditorObjectSelection.cpp]: Always use braces.

File: src/openrct2/windows/EditorObjectSelection.cpp [hunk 4]
```diff
@@ -1877,3 +1877,41 @@ bool editor_check_object_group_at_least_one_selected(sint32 checkObjectType)
...
+	}
+	unload_unselected_objects();
+
+	if (!selectionFlagsExist)
+	{
+		editor_object_flags_free();
+	}
+
+	return numUnselectedObjects;
+}
```
Comment (Gymnasiast) [inline src/openrct2/windows/EditorObjectSelection.cpp]: Always have a newline at the end of the file.

File: src/openrct2/Editor.h [hunk 5]
```diff
@@ -85,6 +85,12 @@ extern "C"
 
     void editor_open_windows_for_current_step();
 
+    /**
+    * Removes all unused objects from the object selection.
+    * @return The number of removed objects.
+    */
+    sint32 editor_remove_unused_objects();
```
Comment (Gymnasiast) [inline src/openrct2/Editor.h]: This doc-block should be on the actual definition, not on this declaration.

File: src/openrct2/windows/EditorObjectSelection.cpp [hunk 6]
```diff
@@ -1877,3 +1877,47 @@ bool editor_check_object_group_at_least_one_selected(sint32 checkObjectType)
     }
     return false;
 }
+
+sint32 editor_remove_unused_objects()
+{
+    bool selectionFlagsExist = _objectSelectionFlags != nullptr;
+    if (!selectionFlagsExist)
```
Comment (Gymnasiast) [inline src/openrct2/windows/EditorObjectSelection.cpp]: Just fold these two `if`s together.

File: src/openrct2/windows/EditorObjectSelection.cpp [hunk 7]
```diff
@@ -1877,3 +1877,47 @@ bool editor_check_object_group_at_least_one_selected(sint32 checkObjectType)
     }
     return false;
 }
+
+sint32 editor_remove_unused_objects()
+{
+    bool selectionFlagsExist = _objectSelectionFlags != nullptr;
```
Comment (Gymnasiast) [inline src/openrct2/windows/EditorObjectSelection.cpp]: `bool selectionFlagsExist = (_objectSelectionFlags != nullptr);`

File: src/openrct2/windows/EditorObjectSelection.cpp (hunk 6 above)
Comment (oli414) [reply src/openrct2/windows/EditorObjectSelection.cpp]: Unfortunately, I can't. I changed the name now though. Should be clearer why it is necessary now.

**Comment by oli414** [comment]
Thanks. That was a bit sloppy of me. I hope it's okay now. Let me know if there's anything else

File: src/openrct2/windows/EditorObjectSelection.cpp (hunk 6 above)
Comment (Gymnasiast) [reply src/openrct2/windows/EditorObjectSelection.cpp]: You can:
```
if (createSelectionFlags && !sub_6AB211())
{
    return 0;
}
```

File: src/openrct2/Editor.h (hunk 5 above)
Comment (IntelOrca) [reply src/openrct2/Editor.h]: No, this is right. It has been done else where, see a lot of the core header files and it also says so in the code style wiki page.

Comment (Gymnasiast) [reply src/openrct2/Editor.h]: Oh, ok. Sorry about that, @oli414 .

File: src/openrct2/windows/EditorObjectSelection.cpp (hunk 6 above)
Comment (oli414) [reply src/openrct2/windows/EditorObjectSelection.cpp]: Overlooked and misunderstood that. Sorry about that.

File: src/openrct2/Editor.h [hunk 8]
```diff
@@ -85,6 +85,12 @@ extern "C"
 
     void editor_open_windows_for_current_step();
 
+    /**
```
Comment (Gymnasiast) [inline src/openrct2/Editor.h]: I have been informed that in fact, doc-block to go above the declaration and not the definition. In other words, what you originally did was correct. Sorry about that.

Comment (oli414) [reply src/openrct2/Editor.h]: Which means that it is correct now right?

Comment (Gymnasiast) [reply src/openrct2/Editor.h]: Er, yes.

**Comment by Gymnasiast** [comment]
Good, it's merged. I think the NE crowd will like it. 👍

**Pull request merged**